### ISSUE TITLE
Webhook triggered on Order Cycle close

### ISSUE THREAD

**Issue by lin-d-hop** [openfoodfoundation/openfoodnetwork#10528]
## Description

As an API user I want to be notified when an OC closes.
As an API user I want to be able to turn notifications on and off

## Acceptance Criteria & Tests

1. We need to be able to turn webhook posting on and off via the UI. API settings are at the User level in Developer Settings.
The UI will exist in https://openfoodnetwork.org.uk/account#/developer_settings
This UI is visible when super-admin enables 'API Key View' for the user.
2. This setting will be called 'Post webhook on Order Cycle Closed'. In line with other settings two radio buttons will denote enabled/disabled:
![Screenshot from 2022-08-31 17-05-57](https://user-images.githubusercontent.com/8669848/187726241-693cd95f-8b03-4409-bec6-e4e76b47d76d.png)
3. When 'enabled' a url is shown where the payload will be sent to.
4. The webhook payload will be triggered when an OC opens for any enterprises this user _manages_ and _owns_. It is applicable for all _distributor_ and _coordinator_ enterprises (and not for one's for which the user _supplies_ only).
5. A tooltip should be included saying 'This will trigger a webhook to post the following payload to the given URL when an order cycle closes:
  "id": STRING,
  "at": DATETIME,
  "type": "order_cycle_close",
  "data": {
    "id": INTEGER,
    "name": STRING,
    "open_at": DATETIME,
    "close_at": DATETIME,
    "enterprise_id": INTEGER,
    "enterprise_name": STRING
    "tags": [STRING]
  }"

6. The payload delivered should be:
  "id": STRING (ID of the webhook)
  "at": DATETIME (when payload sent)
  "type": "order_cycle_close",
  "data": {
    "id": INTEGER (ID of the OC)
    "name": STRING (name of the OC)
    "open_at": DATETIME (OC opening time)
    "close_at": DATETIME (OC closing time)
    "enterprise_id": INTEGER (enterprise ID in DB)
    "enterprise_name": STRING (enterprise name)
    "tags": [STRING] (tags on the OC)
  }

For example:
{
  "id": "EV09374580",
  "at": "2022-06-15T04:18:00-04:00",
  "type": "order_cycle_close",
  "data": {
    "id": 79234596,
    "name": "Aug 3rd"
    "open_at": "2022-06-15T04:18:00-04:00",
    "close_at": "2022-07-15T04:18:00-04:00",
    "enterprise_id": 2093846
    "enterprise_name": "Fictional Hub"
  }
  }

6. Unit Test coverage as standard
7. Translations as standard on the Enterprise Settings page.

## Other useful Bits
This tutorial looks good:
https://keygen.sh/blog/how-to-build-a-webhook-system-in-rails-using-sidekiq/

**Comment by lin-d-hop** [comment]
This is basically a copy paste of OC Open, but on reading through and adjusting I think it is all relevant. Let me know if anything doesn't make sense or if you have any questions.

**Comment by RachL** [comment]
@openfoodfoundation/developers @openfoodfoundation/testers is this one a papercut size or bigger?

**Comment by dacook** [comment]
I think this could be bigger, because #9616 only includes provision for _one_ webhook, so it won't be as simple as creating one more event type. I'll estimate:

1. ~~Add event type field to webhook endpoints - **1hr**~~
2. ~~Add event type to user account form - up to **1day**~~ Done in #13777 
   - TBC: can a user add more than one of each type?
4. [tech debt] Rename existing OrderCycleClosingJob to OrderCycleClosedJob -  **1hr**
    - it requires a safe transition for existing enqueud jobs
     (we chose the past tense for OrderCycleOpenedJob because it doesn't perform the opening, it happens after it has opened. We'd like to bring this existing job name in line with that. [refer](https://github.com/openfoodfoundation/openfoodnetwork/pull/9687#discussion_r990857156) )
5. [fix] Change `at` to ISO date format (this was missed in the initial implementation) - **10min**
6. Trigger webhook when order cycle closes (from newly-renamed OrderCycleClosedJob) - **1hr**

**Comment by dacook** [comment]
Question: can a user add more than one of each type of webhook? (eg add Order Cycle Closed twice, once to Zapier and once to N8N).

I think yes, it should be allowed, to make life easier. Sure, you could probably re-post from one service to the next, but this might cost more or be a challenge for a non-technical user. There's a chance that someone could abuse it and add _heaps_ of endpoints, but I'm sure there's plenty of worse ways to abuse the system (run a report for 12months of data.. :trollface: )

A preview of #9616 for reference:
![](https://user-images.githubusercontent.com/4188088/223575770-879ca807-e87d-428c-b5e5-36895c7d7439.png)

**Comment by mkllnk** [comment]
> Rename existing OrderCycleClosingJob to OrderCycleClosedJob

You need compatibility for enqueued jobs to be found, right? You can always keep the old class for compatibility and use the new code:
```rb
# Temporary class for compatibility.
# Can be removed ten minutes after this change has been deployed.
class OrderCycleClosingJob < OrderCycleClosedJob; end
```

**Comment by mkllnk** [comment]
> Add event type to user account form - up to 1day

Yes, the UI changes are the biggest part. But can we do it incrementally? Maybe we just duplicate what we have and have one field for `opened` and one field for `closed`. Supporting multiple webhooks would be good but can be another issue.

**Comment by audez** [comment]
Hi! Thanks for this <3 
Will we be able to activate the feature as a superadmin? 

Indeed, in FR we will mainly use this webhook to trigger sending a newsletter to a user's client at the end of each OC, so we need to enable the webhook for this user.

**Comment by dacook** [comment]
You're right, it will need to be enabled for the user (that is, the owner of the coordinator/distributor enterprise). 
As it is currently, you need to log in as that user to access the Account > Developer Settings section. So no, you won't be able to activate it as a superadmin via Admin > Users > Edit.
Hopefully you could give them the instructions on how to enable it themselves.

**Comment by audez** [comment]
ok thanks! it will be another step to ask them (we already ask to create an account in a newsletter tool and send us the API key) but I guess they can live with it :)

**Comment by mkllnk** [comment]
> papercut (from memory this means under 1 day development time)

I think that's our Fair Food definition. OFN uses a half day for a Ri dev (experienced). That can be a day for Shu and Ha devs.

**Comment by mkllnk** [comment]
Here is the last definition: https://community.openfoodnetwork.org/t/papercuts-revised-15-oct-2020/1765

I think that we are a bit relaxed with papercuts. Most papercuts take more than 5 minutes to discuss. :wink:
